Encapsulate Vert OAuth2 code inside a single class

sberyozkin opened the pull request:
This PR is related to #14492 but does not do the actual migration, what it does is:
- encapsulates all the Vert.x Oauth2 code inside a single new class `OidcRuntimeClient`.

IMHO it is a good refactoring on its own because right now the use of Vertx OAuth2 is not-coordinated across the quarkus-oidc code and we have some nested/complex code with some casting added on top of it.

So in this PR I've tried to simplify the code along the way to use Uni in a more structured way; also I use now the existing `AuthorizationCodeTokens` holder to pass the 3 tokens around instead of dealing with Vertx Oauth2 `AccessToken` result which has both raw and verified tokens - either locally or remotely - this is also why `TokenVerificationResult` is introduced - going forward I'd rather avoid the introspection completely if the local JWK can be refreshed - now we can't avoid it.

Also during the code flow we use Vert.x OAuth2 to build a redirect URI - I refactored it and made it local, otherwise it is very expensive - the properties are passed as JsonObject which is then copied internally etc.

So for now this PR is only a clean up PR - IMHO it makes sense to stress the updated code a bit - even though nothing changed fundamentally - the code has changed significantly due to the Uni related updates.

It will then also be easier to see the actual migration updates because most of them will actually be local to the new `OidcRuntimeClient` class - and if I hit some issues then it will be still easier to migrate to Vert.x 4 APIs

CC @cescoffier

sberyozkin committed:
Encapsulate Vert OAuth2 code inside a single class

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:
```diff
@@ -268,102 +284,82 @@ private boolean shouldAutoRedirect(TenantConfigContext configContext, RoutingCon
+                                    context.put(NEW_AUTHENTICATION, Boolean.TRUE);
+                                    context.put(CODE_FLOW_ACCESS_TOKEN, tokens.getAccessToken());
+                                    authenticate(identityProviderManager, new IdTokenCredential(tokens.getIdToken(), context))
+                                            .subscribe().with(new Consumer<SecurityIdentity>() {
```
I believe we can simplify this a bit, but that's not the focus of this PR.

sberyozkin replied:
@cescoffier That is a good point - I've refactored it the same way the re-authentication is done a few blocks above, with `.map` and `transform` for the exception, and it looks much better, going to push a bit later

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRuntimeClient.java:
```diff
@@ -0,0 +1,238 @@
+                    @Override
+                    public void handle(AsyncResult<AccessToken> event) {
+                        if (event.failed()) {
+                            uniEmitter.fail(new AuthenticationFailedException(event.cause()));
```
you need to `return` here or add a `else` clause.

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRuntimeClient.java:
```diff
@@ -0,0 +1,238 @@
+        tokenImpl.userInfo(new Handler<AsyncResult<JsonObject>>() {
+            @Override
+            public void handle(AsyncResult<JsonObject> event) {
+                if (event.failed()) {
```
There are plenty of async result <-> emitter code in this class. We should extract that into a specific method and maybe optimize (emitter is not great for this).

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRuntimeClient.java:
```diff
@@ -0,0 +1,238 @@
+                    }
+                });
+            }
+        }).await().indefinitely();
```
That will be problematic

1. it may block forever
2. you may block the wrong thread 

Is it possible to return a `Uni<OidcRuntimeClient>` ?

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRuntimeClient.java:
```diff
@@ -0,0 +1,238 @@
+                        uniEmitter.complete(createClient(auth, oidcConfig));
+                    });
+                }
+            }).await().indefinitely();
```
Same comment here.

sberyozkin commented:
@cescoffier thanks for the review so far, I'll have a look tomorrow. I think those `await().indefinitely()` methods are only used at the bootstrap time, I certainly did not add them with this PR :-), the default (statically configured) tenants should be available by the time the synthetic bean is returned from the recorder - while inside that await block the connection is retried a few times and either succeeds or fails. But for the dynamically created tenants, those which are created at some request time, are definitely created async :-).
I'll check tomorrow

cescoffier commented:
@sberyozkin Great! I didn't catch the bootstrap time. In that case, it's safe.

sberyozkin commented:
Hi Clement @cescoffier 

I've resolved 2 of your comments (simplify the authentication Uni sequence in `CodeAuthenticationMechanism`) and the one about missing `else` in the async callbacks - please check if they are fixed from your point of view.

Re the `await().indefinitely()` - unfortunately I've confirmed it is also currently used on the dynamic path (where we create the connection during some request) - but it is called at the higher level from a Uni scoped blocking executor.

For this specific case I'd like to make sure `Uni<OidcRuntimeClient>` is returned as you suggested - but it would be a complex enough/sensitive update - especially given that the retries are currently done to reconnect in case of the failures. This PR only encapsulated this code, so at least it is not to blame for it :-). But I've added a 3 sec timeout instead of `.indefinitely()` - that will protect against some accidental blocking.  

Likewise, re your other comment that you see too many emitters in `OidcRuntimeClient` :-) - I've just moved some existing code there and I think I also managed to drop one (or may be even two) emitter(s) in `OidcIdentityProvider` as part of the process. All of this code will be replaced in the next phase with the migration to `Vert.x Oidc Web Client` where we'd get `Uni` coming directly from it so the code will be looking very different. This is why I'd propose to refactor all the emiitters/optimize in the final PR resolving the migration issue.

Are you OK with it ?
Thanks

sberyozkin commented:
Hi @cescoffier I've followed up with another update related to these timeouts. By the way I've checked the vertx options, the default is 60 secs - long - but the connection request will never time out :-).

So I've added a `connection-timeout` property, default is 10 secs (passed to the client options, similarly how to Vincent, Georgios do it). This for a single connection request. `quarkus-oidc` can retry - for example, if another  `connection-delay` property is set to 20 secs then it will try reconnecting 10 times with a 2 secs delay. So the total connection delay in this case would be around `20 + 10*10` seconds - this is what I now set `Uni.await()` to (plus 3 secs just in case).
I think we are totally safe here, when it comes to the possibility of Uni blocking at the connection time :-), for now - I've resolved the 2 related comments - reopen please if you don't agree.

So the only remaining comment for now - is `OidcRuntimeClient` optimization to do with the emitters - but as I said this code will be totally different in a follow up PR once I switch to the Vertx Mutiny WebClient

Thanks

cescoffier reviewed (approved):
LGTM!

sberyozkin commented:
@cescoffier thanks Clement, we'll be trying improving the `Uni-awareness` of this code :-) 
@pedroigor Hey Pedro - can you please do a quick check - as described above this mainly encapsulates the Vertx OAuth2 code with a few simplifications/optimizations along the way

sberyozkin commented:
@pedroigor thanks
